File: .dev/orchestrator/commands/verify-flow.js
import fs from 'fs/promises'
import path from 'path'
import chalk from 'chalk'
import matter from 'gray-matter'

export async function verifyFlow(specPath) {
  try {
    console.log(chalk.cyan('\n--- Zuri: Verify Feature Spec ---\n'))

    // Resolve spec path relative to cwd
    const fullPath = path.isAbsolute(specPath)
      ? specPath
      : path.join(process.cwd(), specPath)

    let content
    try {
      content = await fs.readFile(fullPath, 'utf-8')
    } catch {
      console.error(chalk.red(`  File not found: ${specPath}`))
      console.log(chalk.dim('  Hint: provide path relative to project root, e.g. docs/product/features/my-feature.md'))
      process.exitCode = 1
      return
    }

    // Parse frontmatter
    let parsed
    try {
      parsed = matter(content)
    } catch {
      console.error(chalk.red('  Failed to parse frontmatter'))
      process.exitCode = 1
      return
    }

    const { data: fm } = parsed
    const body = parsed.content

    const results = []
    let hasCriticalFail = false

    // --- Check 1: Frontmatter required fields ---
    const requiredFields = ['status', 'module', 'priority', 'author']
    const missingFields = requiredFields.filter((f) => !fm[f] || fm[f] === '')
    if (missingFields.length === 0) {
      results.push({ label: 'Frontmatter', status: 'pass', detail: 'All required fields present' })
    } else {
      results.push({
        label: 'Frontmatter',
        status: 'fail',
        detail: `Missing: ${missingFields.join(', ')}`,
      })
      hasCriticalFail = true
    }

    // --- Check 2: Data Flow section ---
    if (/## 3\. Data Flow|## Data Flow/i.test(body)) {
      results.push({ label: 'Data Flow', status: 'pass', detail: 'Section found' })
    } else {
      results.push({ label: 'Data Flow', status: 'fail', detail: 'Missing' })
      hasCriticalFail = true
    }

    // --- Check 3: API Endpoints section ---
    if (/## 4\. API Endpoints|## API Endpoints/i.test(body)) {
      // Count endpoint rows (lines starting with | GET/POST/PUT/PATCH/DELETE)
      const endpointLines = body
        .split('\n')
        .filter((l) => /^\|\s*(GET|POST|PUT|PATCH|DELETE)\s/i.test(l))
      const detail =
        endpointLines.length > 0
          ? `${endpointLines.length} endpoint(s) defined`
          : 'Section found (no endpoints listed yet)'
      results.push({ label: 'API Endpoints', status: 'pass', detail })
    } else {
      results.push({ label: 'API Endpoints', status: 'fail', detail: 'Missing' })
      hasCriticalFail = true
    }

    // --- Check 4: Roles & Permissions section ---
    if (/## 6\. Roles|## Roles\s*[&|and]*\s*Permissions/i.test(body)) {
      results.push({ label: 'Roles & Permissions', status: 'pass', detail: 'Section found' })
    } else {
      results.push({
        label: 'Roles & Permissions',
        status: 'warn',
        detail: 'No roles section defined',
      })
    }

    // --- Check 5: Schema changes → ADR reference ---
    const mentionsSchema =
      /schema|prisma|migration|model\s+\w+/i.test(body)
    if (mentionsSchema) {
      const adrRef = fm.adr
      const hasAdr =
        adrRef && ((Array.isArray(adrRef) && adrRef.length > 0) || (typeof adrRef === 'string' && adrRef.trim()))
      if (hasAdr) {
        const adrList = Array.isArray(adrRef) ? adrRef.join(', ') : adrRef
        results.push({
          label: 'Schema / ADR',
          status: 'pass',
          detail: `ADR referenced: ${adrList}`,
        })
      } else {
        results.push({
          label: 'Schema / ADR',
          status: 'fail',
          detail: 'Schema changes mentioned but no ADR referenced in frontmatter',
        })
        hasCriticalFail = true
      }
    } else {
      results.push({
        label: 'Schema / ADR',
        status: 'pass',
        detail: 'No schema changes detected',
      })
    }

    // --- Check 6: Status = APPROVED ---
    if (fm.status === 'APPROVED') {
      results.push({ label: 'Status', status: 'pass', detail: 'APPROVED' })
    } else if (fm.status === 'REVIEW') {
      results.push({
        label: 'Status',
        status: 'warn',
        detail: `${fm.status} (needs approval)`,
      })
    } else {
      results.push({
        label: 'Status',
        status: 'fail',
        detail: `${fm.status || 'UNKNOWN'} (needs approval)`,
      })
      hasCriticalFail = true
    }

    // --- Print results ---
    const icon = { pass: chalk.green('PASS'), fail: chalk.red('FAIL'), warn: chalk.yellow('WARN') }
    const sym = { pass: chalk.green('  +'), fail: chalk.red('  x'), warn: chalk.yellow('  !') }

    const overallPass = !hasCriticalFail
    console.log(
      overallPass
        ? chalk.green.bold('  PASS: Ready to implement')
        : chalk.red.bold('  FAIL: Not ready to implement')
    )
    console.log()

    for (const r of results) {
      console.log(`${sym[r.status]} ${r.label}: ${icon[r.status]} ${chalk.dim(r.detail)}`)
    }

    console.log()
    process.exitCode = overallPass ? 0 : 1
  } catch (error) {
    console.error(chalk.red('[verify-flow]'), error.message)
    process.exitCode = 1
  }
}

File: scripts/verify-flow.py
#!/usr/bin/env python3
"""
Zuri: Verify Feature Spec — ตรวจ spec ครบก่อน implement
Usage: python scripts/verify-flow.py docs/product/specs/FEAT05-crm.md
"""
import re
import sys
from pathlib import Path

ROOT = Path(__file__).parent.parent


def parse_frontmatter(content: str) -> tuple[dict, str]:
    """Return (frontmatter_dict, body_text)."""
    m = re.match(r'^---\n(.*?)\n---\n?(.*)', content, re.DOTALL)
    if not m:
        return {}, content
    fm = {}
    for line in m.group(1).splitlines():
        if ':' in line:
            k, _, v = line.partition(':')
            v = v.strip().strip('"').strip("'")
            # handle list values: [a, b]
            if v.startswith('['):
                v = [x.strip().strip('"') for x in v.strip('[]').split(',') if x.strip()]
            fm[k.strip()] = v
    return fm, m.group(2)


def check(label: str, status: str, detail: str) -> dict:
    return {"label": label, "status": status, "detail": detail}


def main():
    if len(sys.argv) < 2:
        print("Usage: python scripts/verify-flow.py <spec-path>")
        print("  e.g. python scripts/verify-flow.py docs/product/specs/FEAT05-crm.md")
        sys.exit(1)

    spec_path = Path(sys.argv[1])
    if not spec_path.is_absolute():
        spec_path = ROOT / spec_path

    print("\n--- Zuri: Verify Feature Spec ---\n")

    if not spec_path.exists():
        print(f"  File not found: {sys.argv[1]}")
        print("  Hint: path relative to project root, e.g. docs/product/specs/FEAT05-crm.md")
        sys.exit(1)

    content = spec_path.read_text(encoding="utf-8")
    fm, body = parse_frontmatter(content)

    results = []
    critical_fail = False

    # --- Check 1: Status + Author fields ---
    has_status = fm.get("status") or re.search(r"\*\*Status:\*\*\s*\w+", body)
    has_author = fm.get("author") or re.search(r"\*\*Author:\*\*", body)
    missing = []
    if not has_status:
        missing.append("Status")
    if not has_author:
        missing.append("Author")
    if not missing:
        results.append(check("Frontmatter", "pass", "All required fields present"))
    else:
        results.append(check("Frontmatter", "fail", f"Missing: {', '.join(missing)}"))
        critical_fail = True

    # --- Check 2: Data Flow section ---
    if re.search(r"## 4\. Data Flow|## Data Flow", body, re.IGNORECASE):
        results.append(check("Data Flow", "pass", "Section found"))
    else:
        results.append(check("Data Flow", "fail", "Missing ## Data Flow section"))
        critical_fail = True

    # --- Check 3: Feature Breakdown / Overview ---
    if re.search(r"## [13]\. (Feature Breakdown|Overview)", body, re.IGNORECASE):
        results.append(check("Feature Breakdown / Overview", "pass", "Section found"))
    else:
        results.append(check("Feature Breakdown / Overview", "fail", "Missing section"))
        critical_fail = True

    # --- Check 4: Roles & Permissions ---
    if re.search(r"## 5\. Roles|## Roles\s*[&and]*\s*Permissions", body, re.IGNORECASE):
        results.append(check("Roles & Permissions", "pass", "Section found"))
    else:
        results.append(check("Roles & Permissions", "warn", "No roles section defined"))

    # --- Check 5: Schema changes → ADR required ---
    mentions_schema = bool(re.search(r"schema|prisma|migration|model\s+\w+", body, re.IGNORECASE))
    if mentions_schema:
        adr_ref = fm.get("adr")
        has_adr = bool(adr_ref and (
            (isinstance(adr_ref, list) and adr_ref) or
            (isinstance(adr_ref, str) and adr_ref.strip())
        ))
        if has_adr:
            adr_list = ", ".join(adr_ref) if isinstance(adr_ref, list) else adr_ref
            results.append(check("Schema / ADR", "pass", f"ADR referenced: {adr_list}"))
        else:
            results.append(check("Schema / ADR", "fail",
                                 "Schema changes mentioned but no ADR in frontmatter"))
            critical_fail = True
    else:
        results.append(check("Schema / ADR", "pass", "No schema changes detected"))

    # --- Check 6: Status = APPROVED ---
    body_status = re.search(r"\*\*Status:\*\*\s*(\w+)", body)
    spec_status = fm.get("status") or (body_status and body_status.group(1)) or "UNKNOWN"
    if spec_status == "APPROVED":
        results.append(check("Status", "pass", "APPROVED"))
    elif spec_status == "REVIEW":
        results.append(check("Status", "warn", f"{spec_status} (needs approval)"))
    else:
        results.append(check("Status", "fail", f"{spec_status} (needs approval)"))
        critical_fail = True

    # --- Print results ---
    sym = {"pass": "+", "fail": "x", "warn": "!"}
    overall = "PASS: Ready to implement" if not critical_fail else "FAIL: Not ready to implement"
    print(f"  {overall}\n")

    for r in results:
        print(f"  {sym[r['status']]} {r['label']}: {r['status'].upper()} — {r['detail']}")

    print()
    sys.exit(0 if not critical_fail else 1)


if __name__ == "__main__":
    main()
